### PR TITLE
More TypeScript

### PR DESCRIPTION
**Pull request by Turbo87**
This PR extracts more TypeScript interfaces and applies them all over the codebase. It also uses them to ensure the correctness of some refactorings, primarily f9cbde0 which replaces the `Concatenator` class with a simpler `concat` function.

### CODE REVIEW

File: test/vlq-tests.ts [hunk 1]
```diff
@@ -109,29 +109,27 @@ describe('test encode', function() {
   it('encoder', function() {
     // (lines + segemnts * 6) = byte_count
     let decoded = {
-      mappings: {
```
Comment (krisselden) [inline test/vlq-tests.ts:112]: Is this white space only? If this is linter related can it be a separate PR? This PR is already not very focused

File: src/vlq.ts [hunk 2]
```diff
@@ -13,11 +15,11 @@ export function encodeVLQ(writer, v) {
     if (cont) {
       digit |= 32;
     }
-    writer.buf[writer.ptr++] = uint6ToASCII[digit];
```
Comment (krisselden) [inline src/vlq.ts:16]: Can you trace this to see that it inlines without issue?

File: src/mappings-decoder.ts [hunk 3]
```diff
@@ -24,22 +25,22 @@ export default class MappingsDecoder {
     this.delegate = delegate;
   }
 
-  decode(reader) {
-    while (reader.ptr < reader.limit) {
-      switch (reader.buf[reader.ptr]) {
+  decode(reader: Reader) {
+    while (reader.hasNext()) {
+      switch (reader.peek()) {
```
Comment (krisselden) [inline src/mappings-decoder.ts:30]: Loops can be weird, especially with switches, please trace in v8 to see how this loop is handled

File: test/vlq-tests.ts (hunk 1 above)
Comment (Turbo87) [inline test/vlq-tests.ts:112]: nope, it's dropping the outer `mappings` property since `MappingsEncoder.encode()` no longer expects that. it only needs the mappings so there is no need for passing it in a wrapped form or as a full sourcemap. (see https://github.com/krisselden/fast-source-map/pull/14/commits/f3dd55cefec17766148af780c6391e8d1a0fac2b)

Comment (Turbo87) [inline test/vlq-tests.ts:112]: > This PR is already not very focused

Sorry about that. If you want I can split it up a bit more. The commits themselves should be reasonably self-contained so you could also review them one by one if that helps with the focus.

**Comment by krisselden** [comment]
Was this benchmarked? Also please follow this https://gist.github.com/krisselden/d3ce3cbb37cc6035b0927fdbf4117b3c and look at the loop

File: src/vlq.ts (hunk 2 above)
Comment (Turbo87) [inline src/vlq.ts:16]: I'll probably need some help on that. How can I test this?

**Comment by Turbo87** [comment]
> Was this benchmarked? Also please follow this https://gist.github.com/krisselden/d3ce3cbb37cc6035b0927fdbf4117b3c and look at the loop

I'll play around with it and let you know what I discover.

**Comment by Turbo87** [comment]
@krisselden your instructions seem to be focused on running/tracing this in Chrome (which explains why this was using rollup to generate something that a browser could load). Is something similar available to trace with Node directly? I think I saw something like that in a Google I/O video recently but I'm not sure if it is actually the same thing or something totally different.

**Comment by krisselden** [comment]
@Turbo87 sorry, the flags in that --js-flags can be passed to node without the other flags, just what is inside the quotes

**Comment by krisselden** [comment]
@Turbo87 I'm just going to merge this now, I can address the other concerns after
